Plugin system for tracking store implementations

zblz opened the pull request:
This PR implements a plugin system for the tracking store implementations. It follows the general principles outlined in the following design document: https://gist.github.com/zblz/9e337a55a7ba73314890be68370fa69a.

External packages would then be able to register an entrypoint against the group `mlflow.tracking_store` that fulfill a given scheme by providing a function that is called with `store_uri` and `artifact_uri` arguments to instantiate a custom implementation of the `AbstractStore` class. It would look like this in the `setup.py` of the `mlflow-foo-plugin` package, where it is defining the `mlflow_foo_plugin:custom_builder` function to be the constructor for store URIs with a `foo` scheme:

```
setup(
    name="mlflow-foo-plugin"
    ...
    entry_points={
        "mlflow.tracking_store": "foo=mlflow_foo_plugin:custom_builder"
    }
)
```

We have already added several tests regarding the handling of the entrypoints in the new class, but an open question is the integration testing of this feature. If we want to test it against an external package within the MLflow codebase we'd have to create a different package that is installed before the tests are run. This could be done within the same repo by having a full package under, e.g., `tests/resources/mlflow-plugin`, and run `pip install tests/resources/mlflow-plugin` before running any tests. However, this is quite clunky and it's not clear what it provides over the mocked out entrypoints in the current tests.

We are also planning on implementing a plugin system for the `ArtifactRepository` implementations, so I think it would be a good idea to defer writing docs on both until we have figured out both implementations.

acroz committed:
WIP: Start adding proposed tracking store registry

mociarain committed:
Add TrackingStoreRegistry and update setup.py

At this stage the legacy mlflow.utils._get_store is a wrapper around the
`tracking_store_registry.get_store` method.

zblz committed:
Make tracking_store_registry private

zblz committed:
Only match registry on scheme

This means we have to make a special case for `store_uri='databricks'`,
which does not strictly have a scheme and would otherwise be handled by
the file store as a relative path.

zblz committed:
Collapse http(s) registrations

mociarain committed:
Add tests for the registry

zblz committed:
Remove unused imports in mlflow/projects/__init__.py

This PR removes a couple of unused imports in
mlflow/projects/__init__.py. In addition to general cleanliness, this is
necessary as we are planning to remove
mlflow.tracking.utils._is_local_uri in an upcoming PR related to a
plugin system for tracking store registration.

zblz committed:
Move databricks URI special case to after env var resolution

zblz committed:
Fix TrackingStoreRegistry tests

zblz committed:
Test that register_entrypoints is called at import time

aarondav commented:
At least one thing that would be provided by having `tests/resources/mlflow-plugin` is a complete, end-to-end example of how to create a plugin internal to the MLflow repo. This is useful documentation in itself, and it helps ensure that the unit tests don't accidentally start testing nothing.

I don't think pip installing an internal plugin prior to running tests is particularly clunky, btw -- seems reasonable. We can mark them with an annotation (like `large`) to avoid them running automatically when someone is just trying to run all the unit tests locally.

aarondav commented on mlflow/tracking/utils.py:
```diff
@@ -164,7 +141,7 @@ def _is_databricks_uri(uri):
 
 
-def _get_file_store(store_uri):
+def _get_file_store(store_uri, artifact_uri):  # pylint: disable=W0613
```
Although these methods are internal to mlflow, and can thus be coversioned with changes to the registry, it may be a good idea to treat them as if they were in an external plugin, for consistency.

I would imagine that if I were writing an external plugin, I might write this something like

```
def _get_file_store(store_uri, **_):
```

This would allow this plugin to work with future versions of mlflow which may pass new arguments (e.g., if we added `store_token` today, it would break historical plugins that only take store_uri and artifact_uri). It also avoids the pylint warning (although `def _get_file_store(store_uri, **kwargs)` is more readable/expected, so may be reasonable to keep that).

Thoughts?

zblz replied:
I've changed it to `**_`.

aarondav commented on mlflow/tracking/utils.py:
```diff
@@ -198,11 +182,60 @@ def get_db_profile_from_uri(uri):
+_tracking_store_registry.register_entrypoints()
+
+
+def _get_store(store_uri=None, artifact_uri=None):
```
As far as I can tell, `artifact_uri` is only provided by `_download_artifact_from_uri` in order to call `ArtifactRepository.from_artifact_uri`. This is something that may be good to remove in a follow-up PR, assuming we can remove `store` from the arguments of `ArtifactRepostiory`.

aarondav commented on mlflow/tracking/utils.py:
```diff
@@ -198,11 +182,60 @@ def get_db_profile_from_uri(uri):
     return RestStore(lambda: get_databricks_host_creds(profile))
 
 
+class TrackingStoreRegistry:
```
Mind adding a brief class comment on what this is responsible for?

zblz replied:
Done, please take a look to see if they are clear enough.

aarondav commented on mlflow/tracking/utils.py:
```diff
@@ -198,11 +182,60 @@ def get_db_profile_from_uri(uri):
+            store_builder = self._registry[scheme]
+        except KeyError:
+            raise MlflowException(
+                "Could not find a registered tracking store for: {}".format(store_uri)
```
We could print the accepted URI schemes for the sake of debuggability.

aarondav commented on tests/tracking/test_utils.py:
```diff
@@ -168,6 +169,62 @@ def test_get_store_databricks_profile():
+    mock_plugin = mock.Mock()
+    tracking_store.register("mock-scheme", mock_plugin)
+    assert "mock-scheme" in tracking_store._registry
+    assert tracking_store.get_store("mock-scheme://fake-host/fake-path") == mock_plugin.return_value
```
May be good to have a test that verifies we invoke the plugin with the appropriate arguments (currently, just store_uri and artifact_uri).

zblz committed:
Use **_ in internal plugin-like functions

zblz committed:
List registered schemes in exception message

zblz committed:
Add minimal test plugin and handle plugin loading failure

In order to avoid breaking mlflow import, we always catch exceptions
raised by `entrypoint.load()` for misconfigured plugins and then raise
them as a warning.

zblz committed:
Test store returned by plugin

zblz committed:
Add docstrings to TrackingStoreRegistry

zblz committed:
Add test for instantiation arguments of plugin

zblz committed:
Simplify mlflow_test_plugin by parsing URI in PluginFileStore

zblz commented:
@aarondav - I've implemented all of your comments above, including adding the test plugin on `tests/resources/mlflow_test_plugin`. It would be great if you could take a look again.

The plugin contains a `PluginFileStore` class that parses the URI and passes the path to a regular `FileStore`, so that the class constructor can be used directly without a `get_file_store` builder as is the case for the `FileStore` right now. In the context of @acroz's [API changes proposal](https://gist.github.com/acroz/48a7b9455f1a054f6b14acf676d6df46) it would definitely be good if we changed internal store classes to be able to be passed directly to the registry (i.e., only take `store_uri` and `artifact_uri` in init method).

I'll implement some of the common changes to the artifact repository plugin system as well.

zblz committed:
Add test for getting unknown scheme

zblz committed:
Merge remote-tracking branch 'upstream/master' into tracking-store-registry

zblz commented on tests/tracking/test_utils.py:
```diff
@@ -168,6 +169,115 @@ def test_get_store_databricks_profile():
+        plugin_file_store = mlflow.tracking.utils._get_store()
+        assert isinstance(plugin_file_store, PluginFileStore)
+        assert os.path.abspath(plugin_file_store.root_directory) == os.path.abspath("test-path")
+        assert os.path.abspath(plugin_file_store.artifact_root_uri) == os.path.abspath("test-path")
```
Good idea, I'll just add another attribute as part of `PluginFileStore` that is specific to that plugin.

zblz committed:
Test whether it's a plugin rather than FileStore attributes

zblz committed:
Merge remote-tracking branch 'upstream/master' into tracking-store-registry

aarondav commented:
LGTM!

zblz commented:
@aarondav CI is failing on an unrelated Keras test. Could you restart the Python2 build?

aarondav commented:
Merged!
